File: LogiX/Simulation/Simulator.cs
using LogiX.Circuits;
using LogiX.Circuits.Drawables;
using LogiX.Utils;
using Newtonsoft.Json;
using Raylib_cs;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;

namespace LogiX.Simulation
{
    [JsonObject(MemberSerialization.OptOut)]
    class Simulator
    {
        public List<DrawableWire> AllWires { get; set; }
        public List<DrawableComponent> AllComponents { get; set; }

        [JsonIgnore]
        public List<DrawableComponent> SelectedComponents { get; set; }
        [JsonIgnore]
        public List<DrawableComponent> Interactables
        {
            get
            {
                return AllComponents.Where(x => x.GetType().GetInterfaces().Contains(typeof(IUpdateable))).ToList();
            }
        }

        public Simulator()
        {
            AllComponents = new List<DrawableComponent>();
            SelectedComponents = new List<DrawableComponent>();
            AllWires = new List<DrawableWire>();
        }

        public void AddWire(DrawableWire wire)
        {
            AllWires.Add(wire);
        }

        public void RemoveWire(DrawableWire wire)
        {
            if(AllWires.Contains(wire))
                AllWires.Remove(wire);
        }

        public void SelectComponentsInRectangle(Rectangle rec)
        {
            ClearSelectedComponents();
            foreach(DrawableComponent dc in AllComponents)
            {
                if (Raylib.CheckCollisionRecs(rec, dc.Box))
                {
                    AddSelectedComponent(dc);
                }
            }
        }

        public void AddSelectedComponent(DrawableComponent dc)
        {
            if(!SelectedComponents.Contains(dc))
                SelectedComponents.Add(dc);
        }

        public void ClearSelectedComponents()
        {
            SelectedComponents.Clear();
        }

        public void DeleteSelectedComponents()
        {
            foreach(DrawableComponent dc in SelectedComponents)
            {
                foreach(CircuitInput ci in dc.Inputs)
                {
                    if (ci.Signal != null)
                    {
                        DrawableWire dw = (DrawableWire)ci.Signal;
                        dw.From.RemoveOutputWire(dw.FromIndex, dw);
                        RemoveWire((DrawableWire)ci.Signal);
                    }
                }
                foreach (CircuitOutput co in dc.Outputs)
                {
                    co.Signals.ForEach(x =>
                    {
                        DrawableWire dw = (DrawableWire)x;
                        dw.To.RemoveInputWire(dw.ToIndex);
                        RemoveWire((DrawableWire)x);
                    });
                }

                AllComponents.Remove(dc);
            }
            SelectedComponents.Clear();
        }

        public void AddComponent(DrawableComponent dc)
        {
            AllComponents.Add(dc);
        }

        public void RemoveComponent(DrawableComponent dc)
        {
            AllComponents.Remove(dc);
        }

        public DrawableComponent GetComponentFromPosition(Vector2 position)
        {
            foreach(DrawableComponent dc in AllComponents)
            {
                if (Raylib.CheckCollisionPointRec(position, dc.Box))
                {
                    return dc;
                }
            }
            return null;
        }

        public Tuple<int, DrawableComponent> GetComponentAndInputFromPos(Vector2 position)
        {
            foreach(DrawableComponent dc in AllComponents)
            {
                int index = dc.GetInputIndexFromPosition(position);

                if (index != -1)
                {
                    return new Tuple<int, DrawableComponent>(index, dc);
                }
            }
            return null;
        }

        public Tuple<int, DrawableComponent> GetComponentAndOutputFromPos(Vector2 position)
        {
            foreach (DrawableComponent dc in AllComponents)
            {
                int index = dc.GetOutputIndexFromPosition(position);

                if (index != -1)
                {
                    return new Tuple<int, DrawableComponent>(index, dc);
                }
            }
            return null;
        }

        public void Update(Vector2 mousePosInWorld)
        {
            foreach (DrawableComponent dc in Interactables)
            {
                ((IUpdateable)dc).Update(this, mousePosInWorld);
            }
        }

        public void UpdateLogic(Vector2 mousePosInWorld)
        {
            foreach(DrawableComponent dc in AllComponents)
            {
                dc.UpdateInputsAndLogic();
            }

            foreach (DrawableComponent dc in AllComponents)
            {
                dc.UpdateOutputs();
            }
        }

        public void Render(Vector2 mousePosInWorld)
        {
            foreach(DrawableWire wire in AllWires)
            {
                wire.Draw();
            }

            foreach (DrawableComponent dc in AllComponents)
            {
                dc.Draw(mousePosInWorld);
            }
            
            foreach(DrawableComponent dc in SelectedComponents)
            {
                dc.DrawSelected();
            }
        }
    }
}

File: src/simulation/simulator.cpp
#include "simulation/simulator.hpp"
#include "raylib-cpp/raylib-cpp.hpp"

void Simulator::Simulate() {
    for (int i = 0; i < allComponents.size(); i++) {
        allComponents.at(i)->UpdateInputsAndPerformLogic();
    }
    for (int i = 0; i < allComponents.size(); i++) {
        allComponents.at(i)->UpdateOutputs();
    }
}

void Simulator::Update(Vector2 mousePosInWorld) {
    for (int i = 0; i < allComponents.size(); i++)
    {
        allComponents.at(i)->Update(mousePosInWorld);
    }
}

void Simulator::Draw(Vector2 mousePosInWorld) {
    for (int i = 0; i < allWires.size(); i++) {
        allWires.at(i)->Draw();
    }

    for (int i = 0; i < allComponents.size(); i++) {
        allComponents.at(i)->Draw(mousePosInWorld);
    }

    for (int i = 0; i < selectedComponents.size(); i++) {
        selectedComponents.at(i)->DrawSelected();
    }
}


void Simulator::MoveAllSelectedComponents(Vector2 vec) {
    for (int i = 0; i < this->selectedComponents.size(); i++) {
        DrawableComponent* dc = this->selectedComponents.at(i);
        dc->position = dc->position + vec;
    }
}

DrawableComponent* Simulator::GetComponentFromPosition(Vector2 position) {
    for (int i = 0; i < this->allComponents.size(); i++) {
        DrawableComponent* dc = this->allComponents.at(i);
        if (CheckCollisionPointRec(position, dc->box)) {
            return dc;
        }
    }
    return NULL;
}

void Simulator::SelectAllComponentsInRectangle(Rectangle rec) {
    this->ClearSelection();
    for (int i = 0; i < allComponents.size(); i++) {
        if(CheckCollisionRecs(rec, allComponents.at(i)->box)) {
            this->SelectComponent(allComponents.at(i));
        }
    }
}

CircuitIODesc* Simulator::GetComponentInputIODescFromPos(Vector2 position) {
    for (int i = 0; i < this->allComponents.size(); i++)
    {
        DrawableComponent* dc = allComponents.at(i);
        int index = dc->GetInputIndexFromPosition(position);

        if(index != -1) {
            return new CircuitIODesc{true, dc, index};
        }
    }
    return NULL;
}

CircuitIODesc* Simulator::GetComponentOutputIODescFromPos(Vector2 position) {
    for (int i = 0; i < this->allComponents.size(); i++)
    {
        DrawableComponent* dc = allComponents.at(i);
        int index = dc->GetOutputIndexFromPosition(position);

        if(index != -1) {
            return new CircuitIODesc{false, dc, index};
        }
    }
    return NULL;
}

DrawableWire* Simulator::GetWireFromPosition(Vector2 pos) {
    for (int i = 0; i < allWires.size(); i++)
    {
        if(allWires.at(i)->IsPositionOnLine(pos)) {
            return allWires.at(i);
        }
    }
    return NULL;
}

void Simulator::DeleteSelectedComponents() {
    if (selectedComponents.size() == allComponents.size()) {
        selectedComponents = {};
        allComponents = {};
        allWires = {};
    }

    for (int i = 0; i < this->selectedComponents.size(); i++)
    {
        this->RemoveComponent(selectedComponents.at(i));
    }
    ClearSelection();
}

void Simulator::RemoveComponent(DrawableComponent* component) {
    std::vector<DrawableComponent*> comps;

    for (int i = 0; i < allComponents.size(); i++) {
        if (allComponents.at(i) != component) {
            comps.push_back(allComponents.at(i));
            continue;
        }

        DrawableComponent* dc = allComponents.at(i);
        // TODO: do everything that must be done when deleting a component.
        for (int j = 0; j < dc->inputs.size(); j++)
        {
            CircuitInput* ci = dc->inputs.at(j);

            if(ci->HasSignal()) {
                DrawableWire* wire = (DrawableWire*)(ci->GetSignal());
                wire->from->RemoveOutputWire(wire->fromIndex, wire);
                this->RemoveWire(wire);
            }
        }

        for (int j = 0; j < dc->outputs.size(); j++) {
            CircuitOutput* co = dc->outputs.at(j);

            if(co->HasAnySignal()) {
                for (int k = 0; k < co->GetSignals().size(); k++)
                {
                    DrawableWire* wire = (DrawableWire*)(co->GetSignals().at(k));
                    wire->to->RemoveInputWire(wire->toIndex);
                    this->RemoveWire(wire);
                }          
            }
        }    
    }
    allComponents = comps;
}

